packages/x-date-pickers/src/internals/hooks/useField/useField.utils.ts

```diff
@@ -786,20 +786,38 @@ const transferDateSectionValue = <TDate>(
   }
 };
 
+const reliableSectionModificationOrder: FieldSectionType[] = [
+  'year',
+  'month',
+  'day',
+  'weekDay',
+  'hours',
+  'minutes',
+  'seconds',
+  'meridiem',
+];
+
 export const mergeDateIntoReferenceDate = <TDate>(
   utils: MuiPickersAdapter<TDate>,
   dateToTransferFrom: TDate,
   sections: FieldSectionWithoutPosition[],
   referenceDate: TDate,
   shouldLimitToEditedSections: boolean,
 ) =>
-  sections.reduce((mergedDate, section) => {
-    if (!shouldLimitToEditedSections || section.modified) {
-      return transferDateSectionValue(utils, section, dateToTransferFrom, mergedDate);
-    }
+  // cloning sections before sort to avoid mutating it
+  [...sections]
+    .sort(
+      (a, b) =>
+        reliableSectionModificationOrder.indexOf(a.type) -
+        reliableSectionModificationOrder.indexOf(b.type),
```

Comment:
Not a big fan of using `indexOf` since the data can already be represented as an object mapping type to priority.

By the way, if we add a section type and forget this behavior it will automatically get top priority (`indexOf` returns `-1` in such a case)